proposals/0001-metadata-providers.md

```diff
@@ -0,0 +1,489 @@
+    let logger: Logger
+    var baggage: Baggage?
+
+    init(logger: Logger) {
+        self.logger = logger
+        // in here we can still access the task-local Baggage
+        self.baggage = Baggage.current
+    }
+
+    func channelActive(context: ChannelHandlerContext) {
+        logger.debug("Channel active.", baggage: self.baggage)
+        // debug [trace-id: 42] Channel active.
+    }
+
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        logger.debug("Received request part.", baggage: self.baggage)
+        // debug [trace-id: 42] Received request part.
+    }
+}
+```
+
+> `*` situations where task-local baggage just works include: plain synchronous methods, asynchronous methods, as all forms of structured concurrency (async lets and task groups), as well as un-structured (but _not_ detached) tasks (created by calling `Task { ... }`).
+
+The libraries are then free to continue using their callback or delegate-style APIs, and whenever needing to log or restore the baggage they can, either:
+
+- explicitly pass it to a logger: `log.info("Request finished", metadata: [...], baggage: self.baggage)`
+- restore the contextual task-local baggage and wrap subsequent calls with: `Baggage.current.withValue(self.baggage) { moreThings() }`
+
+### Rendering provider metadata only once in performance-critical code
+
+In some performance-critical code, you may want to render the provider metadata only once. In such cases, you can call the metadata provider explicitly at the point of entry and instead of passing `Baggage` you'd use the `Logger`'s metadata:
+
+```swift
+class PerformanceCriticalRequestStateMachine: ChannelInboundHandler {
+    let logger: Logger
+
+    init(logger: Logger) {
+        var logger = logger
+        // in here we can still access the task-local Baggage
+        if let baggage = Baggage.current, let metadataProvider = logger.metadataProvider {
+            let metadata = metadataProvider(baggage)
+            for (key, value) in metadata {
+                logger[metadataKey: key] = value
+            }
+        }
+        self.logger = logger
+    }
+
+    func channelActive(context: ChannelHandlerContext) {
+        logger.debug("Channel active.", baggage: nil)
+        // debug [trace-id: 42] Channel active.
+    }
+
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        logger.debug("Received request part.", baggage: nil)
+        // debug [trace-id: 42] Received request part.
+    }
+}
+```
+
+Passing `nil` as the `Baggage` will prevent task-local look-ups in these performance sensitive situations.
+
+> NOTE: Most applications shouldn't need to worry about this and just rely on automatic Baggage propagation using task-locals
+
+> WARNING: The context-dependent metadata becomes static which makes it hard to recover the dynamic information back into `Baggage`, e.g. if you wanted to create a child span in one of those callbacks.
+
+## Alternatives considered
```

Comment:
Have you considered defining an abstraction for baggage (without actually using `Baggage`) in swift-log and then have swift-distributed-tracing provide a concrete implementation?

Reply:
that would mean we need another package, to define some "abstract baggage like thing" which then both depend on (somewhat painful, baggage _is_ the "currency" for these things already)... and we'd end up using existentials in swift-log which is likely a bad idea (for performance reasons).

If it's yet again about baggage naming... perhaps we should reconsider the package name, but that already was through so many iterations of naming hell that I'm trying to be careful with that... I don't think we can do "swift-baggage" for example...

Reply:
Though perhaps another option would be to make providers:

```
() -> Metadata 
```

and every provider has to obtain the baggage _inside it_.

Upside: 
 - log does not have to depend on baggage for metadata providers

Downsides:
 - it becomes impossible to manually propagate baggage and libraries _must_ always wrap all code that should set have a trace-id in logs in a task local's `withValue` which has a performance impact -- so the high performance use-cases are hurt by this
 - if multiple metadata providers are used, it is likely they are going to be repeatedly looking up the same baggage task local... in some situations that's fine, in others perhaps somewhat painful (again, for performance reasons)
   - this is possible to work around by carefully designing libraries that offer metadata providers to also accept a `(Baggage) -> Metadata` but that's up to library authors then...

Reply:
generally speaking, I think it is okay to set a somewhat higher bar for library authors. ,and especially authors of core libraries such as async-http-client. would love to understand the exact performance issue with a closure / auto-closure since adding a dependency to swift-log seems like a fairly serious implication.

Reply:
cc @weissi

Reply:
What this design is try to mitigate is tons of thread-local accesses,  since every task-local value "pick up" is a thread-local hit, first and foremost. So if there's multiple hits to the same "contextual metadata container" (the baggage), per single log statement that's adding up. If there's one that's just life -- that's how locals work and we're only ever picking them up when _actually_ going to emit a log statement -- where the formatting of strings and other things absolutely dominate the pickup anyway.

The other reason to have a way to manually pass the baggage, and therefore _must_ have the type known to the library, is that libraries like NIO literarily can't rely on task-locals properly, since the same thread is shared across other callbacks/handlers, and just setting a scoped local works perfectly within swift concurrency, just isn't the right thing in a NIO system -- and picking up the local will also never be right; so, there has to be a way to pass into the same "extract values to log them" using the container manually.

Reply:
As far as benchmarking, we kicked off some initial benchmarks here: https://github.com/slashmo/swift-log/tree/feature/baggage-benchmarks though we need to make sure we're measuring the right things there, I'll look at it before our call tonight a bit more.

Reply:
if we end up deciding we _must_ add an additional container type, I think we should pause and consider where it is best"located". swift-log is designed to be a low level, zero dependencies library and is used across many use cases that do not include distributed tracing, which triggered the above reaction. as such, if we _must_ add such container type, maybe there is a lower level where we can introduce this like the standard library as this is potentially useful outside this immediate context / in other context propagation scenarios. alternatively, we can explore reversing the dependency and add this type to swift-log and have the higher level distributed tracing libraries depend on swift-log. these are just ideas, the point is that we need to explore all options